truss/templates/control/control/helpers/inference_server_controller.py

```diff
@@ -0,0 +1,72 @@
+import os
+import threading
+
+from helpers.inference_server_process_controller import InferenceServerProcessController
+from helpers.patch_applier import PatchApplier
+from helpers.types import Patch
+
+
+class InferenceServerController:
+    """Controls what code the inference server runs with.
+
+    Currently, it only applies locks to various actions and mostly
+    delegates to InferenceServerProcessController.
+    """
+
+    def __init__(
+        self,
+        process_controller: InferenceServerProcessController,
+        patch_applier: PatchApplier,
+        app_logger,
+    ):
+        self._lock = threading.Lock()
+        self._process_controller = process_controller
+        self._patch_applier = patch_applier
+        self._current_running_hash = os.environ.get("HASH_TRUSS", None)
+        self._app_logger = app_logger
+
+    def apply_patch(self, patch_request):
+        with self._lock:
+            req_hash = patch_request["hash"]
+            req_prev_hash = patch_request["prev_hash"]
+
+            self._app_logger.debug(
+                f"request hash: {req_hash}, request prev hash:"
+                f" {req_prev_hash}, current hash: "
+                f"{self._current_running_hash}"
+            )
+
+            if req_hash == self._current_running_hash:
```

Comment:
The idea is that if request hash is the same as currently running hash then that means the caller and the container are in sync wrt truss hash and this is a good point for inference server to be running. Yes, this is part of the inference server startup protocol, inference server is only started once this sync is achieved, otherwise there's the risk of serving the wrong hash.

I'm in two minds about this being here in this function. On the one hand it feels like it is hiding here, on the other hand it fits well here. e.g. ensuring the inference server is running in the truss sync state is meaningful regardless of this being part of the start up protocol. Open to other ideas. 

At the very least it's worth documenting the startup protocol. Let me make sure that it's documented.